diff --git a/src/linux_joystick.c b/src/linux_joystick.c
index 07d41d3..5872f4c 100644
--- a/src/linux_joystick.c
+++ b/src/linux_joystick.c
@@ -27,6 +27,16 @@
 
 #include "internal.h"
 
+#ifndef O_CLOEXEC
+#ifdef __sparc__
+#define O_CLOEXEC      0x400000
+#elif defined(__alpha__) || defined(__hppa__)
+#define O_CLOEXEC      010000000
+#else
+#define O_CLOEXEC      02000000
+#endif
+#endif
+
 #if defined(GLFW_BUILD_LINUX_JOYSTICK)
 
 #include <sys/types.h>
diff --git a/src/posix_time.c b/src/posix_time.c
index a172408..8076cdf 100644
--- a/src/posix_time.c
+++ b/src/posix_time.c
@@ -24,6 +24,7 @@
 //    distribution.
 //
 //========================================================================
+ #define _POSIX_C_SOURCE 199309L
 
 #include "internal.h"
 
diff --git a/src/wl_init.c b/src/wl_init.c
index 3aff476..e98f9c3 100644
--- a/src/wl_init.c
+++ b/src/wl_init.c
@@ -24,6 +24,7 @@
 //
 //========================================================================
 
+#define _POSIX_C_SOURCE 199309L
 #include "internal.h"
 
 #if defined(_GLFW_WAYLAND)
diff --git a/src/wl_window.c b/src/wl_window.c
index 5b491ff..fe19474 100644
--- a/src/wl_window.c
+++ b/src/wl_window.c
@@ -40,7 +40,7 @@
 #include <sys/mman.h>
 #include <sys/timerfd.h>
 #include <poll.h>
-#include <linux/input-event-codes.h>
+#include <linux/input.h>
 
 #include "wayland-client-protocol.h"
 #include "xdg-shell-client-protocol.h"
